### PR TITLE
Code coverage

### PR DESCRIPTION
**Pull request by KyleMcB**
I created two gradle "tasks" it works like gradlew build, except its gradlew codeCoverageReport and gradlew testCodeCoverage
The first creates build/reports/jacoco/codeCoverageReport/html/index.html which is the code coverage report in html form and the second will pass if the number of lines the tests cover is less than %10
and finally created a github action that should runs these automatically and leave a green check or red x next to a pull request indicating if it has passed or not. This branch should have a red X because it has no tests and will be under %10

### CODE REVIEW

**Comment by KyleMcB** [comment]
don't merge this I'm still working out kinks, I just need this pull request open to test github actions

**Comment by KyleMcB** [comment]
OK. tests, code coverage, and github actions are all working as they should! Why are all code coverage tools a nightmare... Anyways. As of right now, there should be a little red x next to this pull request because the code coverage is less than 10%, but I think this branch is ready to be merged into master at this point. Hopefully I can learn enough about mocking that I can write tests for small functions in our classes.

**Comment by KyleMcB** [comment]
Also note: You can download the code coverage report made by the github actions by clicking details, then click artifacts and download the report. Its in html form so double click index.html and you can see a break down of what was and was not covered by the tests

**Review by youn0125 — APPROVED** [review]
I can't understand what all that means, however, I want to merge into master and try out.
Is there a reference that we can study about this?

File: .github/workflows/battlecode.yml [hunk 1]
```diff
@@ -0,0 +1,37 @@
+name: Java CI
```
Comment (KyleMcB) [inline .github/workflows/battlecode.yml:1]: This is a github action workflow file. its kind of like a hybrid between a configuration file and a script that tells on what event the action should be executed on. Then it runs the rest of the script on some github host computer. Every step is self docuemented with its name. Each run is just a shell command

File: build.gradle [hunk 2]
```diff
@@ -8,11 +8,20 @@ apply plugin: 'java'
 
 apply plugin: 'idea'
 apply plugin: 'eclipse'
+apply plugin: 'jacoco'
```
Comment (KyleMcB) [inline build.gradle:11]: jacoco is a java code coverage tool. This line tells gradle to use it.

File: build.gradle [hunk 3]
```diff
@@ -8,11 +8,20 @@ apply plugin: 'java'
...
 
 ext.versions = [
     battlecode: new File(projectDir, "version.txt").text.trim()
     //scala: '2.11.7'
 ]
+jacocoTestReport {
```
Comment (KyleMcB) [inline build.gradle:17]: this section is jacoco's configuration for our project. Although I don't think html.destination even works. I basically punched a configuration file for hours with help from stackoverflow until it worked.

File: build.gradle [hunk 4]
```diff
@@ -322,5 +331,48 @@ task buildMap(type: JavaExec, dependsOn: 'build') {
     classpath = project.sourceSets.main.runtimeClasspath
 }
 
+task codeCoverageReport(type: JacocoReport) {
```
Comment (KyleMcB) [inline build.gradle:356]: This adds the codeCoverageReport command to gradlew. Which will generate the html report in build/report/jacoco/codecoverage/ (that path isn't exact). Again I don't really know anything about gradle or jacoco. I just tried a bunch of different options each times trying to get less errors until it worked.

File: build.gradle [hunk 5]
```diff
@@ -322,5 +331,48 @@ task buildMap(type: JavaExec, dependsOn: 'build') {
...
+    reports {
+        html.enabled true
+    }
+}
+
+codeCoverageReport.dependsOn {
+    subprojects*.test
+}
+
+task testCodeCoverage(type: JacocoCoverageVerification) {
```
Comment (KyleMcB) [inline build.gradle:378]: This adds the gradle command testCodeCoverage which will exit with a status code indicating a pass if the overall coverage is above 10% or return a failure if it is under %10.

File: build.gradle [hunk 6]
```diff
@@ -322,5 +331,48 @@ task buildMap(type: JavaExec, dependsOn: 'build') {
...
+    executionData fileTree(project.rootDir.absolutePath).include("**/build/jacoco/*.exec")
+    sourceSets sourceSets.main
+    subprojects.each {
+        sourceSets it.sourceSets.main
+    }
+    violationRules {
+        rule {
+            limit {
+                minimum = 0.1
```
Comment (KyleMcB) [inline build.gradle:387]: This is where the %10 coverage requirement comes from. If we want to change %10 to %20 we could just change minimum to 0.2

File: gradle.properties [hunk 7]
```diff
@@ -1,5 +1,5 @@
 # modify this file to change project properties
-teamA=examplefuncsplayer
```
Comment (KyleMcB) [inline gradle.properties]: I don't remember why I changed this file in this branch. I probably started working on the playtesting a little bit before I created a new branch.

**Pull request merged**